cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -163,27 +162,41 @@ public void createTethering(FullHttpRequest request, HttpResponder responder, @P
       throw e;
     }
 
-    // We don't need to keep track of the client metadata on the server side.
-    PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaceAllocations(), Collections.emptyMap(),
-                                                 tetherRequest.getDescription());
-    // We don't store the peer endpoint on the server side because the connection is initiated by the client.
-    PeerInfo peerInfo = new PeerInfo(peer, null, TetheringStatus.PENDING, peerMetadata,
-                                     tetherRequest.getRequestTime());
     try {
-      store.addPeer(peerInfo);
-    } catch (PeerAlreadyExistsException pae) {
-      // Peer is already configured, treat this as a no-op.
-      responder.sendStatus(HttpResponseStatus.OK);
-      return;
-    } catch (Exception e) {
+      String content = request.content().toString(StandardCharsets.UTF_8);
+      TetheringConnectionRequest tetherRequest = GSON.fromJson(content, TetheringConnectionRequest.class);
+
+      // We don't need to keep track of the client metadata on the server side.
+      PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaceAllocations(), Collections.emptyMap(),
+                                                   tetherRequest.getDescription());
+      try {
+        PeerInfo existingPeerInfo = store.getPeer(peer);
+        if (existingPeerInfo.getTetheringStatus() == TetheringStatus.REJECTED) {
+          // If peer exists in rejected state, send 403 to the client
+          responder.sendString(HttpResponseStatus.FORBIDDEN, TETHERING_REJECTED_MESSAGE);
+          return;
+        }
+        // Update the peer info. Keep the existing tethering status.
+        PeerInfo updatedPeerInfo = new PeerInfo(peer, null, existingPeerInfo.getTetheringStatus(),
+                                                peerMetadata, tetherRequest.getRequestTime(),
+                                                existingPeerInfo.getLastConnectionTime());
+        store.writePeer(updatedPeerInfo);
+        responder.sendStatus(HttpResponseStatus.OK);
+      } catch (PeerNotFoundException e) {
+        // Peer doesn't already exist, let's add it.
+        PeerInfo peerInfo = new PeerInfo(peer, null, TetheringStatus.PENDING, peerMetadata,
+                                         tetherRequest.getRequestTime());
+        store.writePeer(peerInfo);
+        responder.sendStatus(HttpResponseStatus.OK);
+      }
+    } catch (IOException e) {
       try {
         messagingService.deleteTopic(topicId);
```

Comment:
With this logic, a transient DB error will cause the topic to be deleted. I'm guessing this is here to try to 'roll back' changes we try to add the peer and it fails? However, if there is a problem talking to the DB, there is no way to know if this should be deleted or not.

Instead of blindly creating the topic before interacting with the store, the topic should only be created if the peer was added to the store and it didn't exist before (writePeer() should be changed to return that information). However, in this situation it's now possible to add the peer to the store and then fail to create the topic. So whenever the topic is used, we need to lazily create the topic if it is not found.

Reply:
fixed